### PR TITLE
[Sampling] Allow sampling-mask replay with DisallowedTokensLogitsProcessor

### PR DESCRIPTION
**Pull request by ByronHsu**
## Motivation

Not all custom logit processors are compatible with `return_sampling_mask`. Rule of thumb for a valid custom processor for sampling-mask replay: **do not change the relative probabilities of surviving tokens**. This PR introduces an explicit allowlist containing `DisallowedTokensLogitsProcessor`.

For example, Qwen3.5-4B sets `vocab_size=248320` in [config.json](https://huggingface.co/Qwen/Qwen3.5-4B/blob/main/config.json), but [tokenizer.json](https://huggingface.co/Qwen/Qwen3.5-4B/blob/main/tokenizer.json), including added tokens from [tokenizer_config.json](https://huggingface.co/Qwen/Qwen3.5-4B/blob/main/tokenizer_config.json), defines only **248,077 token IDs**. This leaves **243 unmapped IDs (`248077–248319`)**. Without `DisallowedTokensLogitsProcessor`, the inference engine can sample these unmapped IDs during RL rollouts, producing invalid training data and potentially corrupting policy-gradient updates. Supporting this processor with `return_sampling_mask` lets callers exclude these IDs while preserving correct sampling-mask replay.

For example, with no additional top-k/top-p filtering:

**DisallowedTokensLogitsProcessor (valid)**

```text
                                  hi       you      hey
Original probabilities           50%       30%      20%

DisallowedTokensLogitsProcessor: ban "hi"
Returned mask                     no       yes      yes
Actual probabilities              0%       60%      40%
Replayed probabilities            0%       60%      40%   MATCH
```

**Hypothetical TokenBiasLogitsProcessor (invalid)**

This processor adds `ln(2)` to the logit for `hey`, doubling its probability weight before normalization. The returned mask cannot record this change.

```text
                                  hi       you      hey
Original probabilities           50%       30%      20%

TokenBiasLogitsProcessor: multiply "hey"'s weight by 2
Returned mask                    yes       yes      yes
Actual probabilities            41.7%      25%     33.3%
Replayed probabilities           50%       30%      20%   MISMATCH
```

## Modifications

- Accept `DisallowedTokensLogitsProcessor`, including subclasses that implement hard exclusion, with `return_sampling_mask`.
- Reject unsupported or malformed processors with HTTP 400 during request validation, before they reach the scheduler.
- Check `--enable-custom-logit-processor` before deserializing. Apply the compatibility check only to requests returning sampling masks.

## Accuracy Tests

- 86 unit and sampler tests passed, including hard exclusion in mixed batches on PyTorch and FlashInfer.
- 13 HTTP tests passed on an H200 with `Qwen/Qwen3-0.6B`, covering both sampling backends, excluded tokens and replay log probabilities, rejection/recovery, and deterministic sampling.
- The rejection regression fails on unmodified main and passes with this change.
- Full pre-commit checks passed.

## Speed Tests and Profiling

The change adds request validation only; sampler kernels and per-token execution are unchanged. No speed benchmark was run.

---
### CI States

Latest PR Test (Base): :x: [Run #34088940148](https://github.com/sgl-project/sglang/actions/runs/34088940148)
Latest PR Test (Extra): :hourglass_flowing_sand: [Run #35022705311](https://github.com/sgl-project/sglang/actions/runs/35022705311)
Latest PR Test (AMD ROCm 10): :heavy_minus_sign: **No AMD PR run found for this commit**.

### CODE REVIEW

**Review by nanjiangwill — APPROVED** [review]

**Comment by nanjiangwill** [comment]
good catch!

**Comment by ByronHsu** [comment]
/rerun-test registered/sampling/test_sampling_mask.py registered/unit/managers/test_sampling_mask_validation.py
